hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java

```diff
@@ -130,28 +135,40 @@ private void handlePlatformTransaction(
         try {
             // Setup configuration
             var configuration = configProvider.getConfiguration();
+            final var hederaConfig = configuration.getConfigData(HederaConfig.class);
 
-            final var verifications = getVerifications(state, platformEvent, platformTxn, configuration);
+            final var preHandleResult = getCurrentPreHandleResult(state, platformEvent, platformTxn, configuration);
             recordBuilder.transaction(
-                    verifications.txInfo().transaction(), verifications.txInfo().signedBytes());
-
-            // Read all signature verifications. This will also wait, if validation is still ongoing.
-            final var keyVerifications = new HashMap<Key, SignatureVerification>();
-            for (final var entry : verifications.keyVerifications().entrySet()) {
-                // TODO: Implement timeout
-                final var verification = entry.getValue().get();
-                if (verification.failed()) {
+                    preHandleResult.txInfo().transaction(),
+                    preHandleResult.txInfo().signedBytes());
+
+            // Check all signature verifications. This will also wait, if validation is still ongoing.
+            final var timeout = hederaConfig.workflowVerificationTimeoutMS();
+            final var maxMillis = time.currentTimeMillis() + timeout;
+            final var payerKeyVerification =
+                    preHandleResult.verificationResults().get(preHandleResult.payerKey());
+            if (payerKeyVerification.get(timeout, TimeUnit.MILLISECONDS).failed()) {
+                throw new HandleException(ResponseCodeEnum.INVALID_SIGNATURE);
+            }
+            for (final var key : preHandleResult.requiredKeys()) {
+                final var remainingMillis = maxMillis - time.currentTimeMillis();
+                if (remainingMillis <= 0) {
+                    throw new TimeoutException("Verification of signatures timed out");
```

Comment:
I think this is a problem as it will almost certainly lead the node to ISS. If one node out of the bunch times out for some reason, but the others don't, then they will handle this transaction, but this node won't. That will inevitably lead to an ISS. What is the right thing to do here? Do we block forever? If we do, the node will eventually fall behind and reconnect. Maybe that is a better outcome?

Reply:
Can we recover if we block forever? Is there a facility that kills hanging threads and initiates a reconnect?

Otherwise, I think the right thing to do is flush out all logs and error messages and shut down the node (assuming a monitor on the OS level restarts the node automatically if it is shut down and initiates a reconnect).